### ISSUE TITLE
Problems with prediction speed and accuracy 

### ISSUE THREAD

**Issue by qashqay654** [Microsoft/LightGBM#435]
I'm trying to train and predict lgbm model with my own feval.

First problem is that it takes 45 seconds to prediction of dataset with 100 thousand rows and 500 columns on 50 CPUs. Is it possible to make a prediction faster?

The second one is that prediction in feval function differs from total prediction on the same dataset
```
prediction = gbm.predict(X_te)
feval_prediction = gbm._Booster__inner_predict_buffer[1]
print (feval_prediction == prediction).all()
```
False

How can I evaluate feval with correct prediction?

**Comment by guolinke** [comment]
How did you get the prediction? 
it will be slow if using ```gbm.prediction``` . But using ```feval(pred, dataset)```, it is just some copy ( https://github.com/Microsoft/LightGBM/blob/66b7f03238387951e6a0ba3060353ae098691fbc/src/boosting/gbdt.cpp#L610-L646 ) .

For the differs, it is caused by the multi-threading sum-up order. you should use "almostSame" here.

**Comment by qashqay654** [comment]
@guolinke thank you. And what is the fastest way to get prediction from trees if not using `gbm.prediction` in python? `feval(pred, dataset)` is not working in my case, cause I need to get full vector of predictions.

**Comment by guolinke** [comment]
using ```gbm.prediction``` is always slow. The reason is it needs start from first tree every time you call it. 

What is "full vector" ? the content of```pred``` in ```feval(pred, dataset)```  is same(ignore sum-up issue) as the ```gbm.prediction```.

**Comment by qashqay654** [comment]
Hi, again. 
I've read threads #437 and #437 about difference in val predictions. In my problem I haven't any NaN in data, but still prediction on validation set while training is very different from one I got from predict() function. It doesn't seem to be sum-up error, because sometimes the difference is really huge.
![pred](https://cloud.githubusercontent.com/assets/27771679/25338076/b0680dd0-2906-11e7-9b9f-1136b9d9deba.png)
 Attaching the histogram of two validation predictions.

**Comment by guolinke** [comment]
@qashqay654 
did you use any categorical features ?

**Comment by guolinke** [comment]
@wxchan can you also help for this issue ? maybe something wrong in python-package.

**Comment by qashqay654** [comment]
@guolinke  No, only continuous ones.

**Comment by qashqay654** [comment]
@guolinke also, I coded my own saved model parser and see the same result.

**Comment by guolinke** [comment]
@qashqay654 
OK. 
I need more details. 
1. application,  binary classification  / multi-class or ? 
2. The prediction on all iterations have same issue ? or only when iterations is large ?
3. can you try set num_thread=1 ? 
4. it will be better if you can provide the outputted model file.

**Comment by qashqay654** [comment]
1. it's a regression problem with dart boosting
2. same on 10 and 1000 iterations
3. I've just fitted 10 iterations on 1 CPU, but the problem persists
4. sorry, I can not publicly upload model

**Comment by guolinke** [comment]
@qashqay654 Can you try the ```gbdt```, not ```dart``` ?

**Comment by qashqay654** [comment]
Tried gbdt on 1 CPU with the same result
![pred10_1cpu](https://cloud.githubusercontent.com/assets/27771679/25342185/e259bf02-2913-11e7-9a19-51c1f7462259.png)

**Comment by Laurae2** [comment]
@qashqay654 Can you provide a reproducible example using a demo example for the last issue you have?

**Comment by qashqay654** [comment]
@Laurae2 I can't face with that problem on example dataset. Datasets in ./examples/regression/regression.test gives completely correct predictions on my params

**Comment by guolinke** [comment]
@qashqay654 
Not use any init_predictor / init_score ?

**Comment by qashqay654** [comment]
@guolinke no, I'm just using code from simple_example.py with my own dataset.

**Comment by guolinke** [comment]
@qashqay654 still cannot figure out why. 
can you provide a sampled/transformed data (not including the sensitive information), that can reproduce this problem ? 
Tough you said your data don't contain NaN, can you also try fill NaN by zero (include the labels) explicitly?

**Comment by qashqay654** [comment]
@guolinke I've fitted only one tree and tried to find points where predictions differ. So there is only one split where I see the problem. It has a threshold equals to -12943187 and has two leafs as a children. The value of the feature in these points equals to 0.0, so `prediction()` function works correctly and returns the result of the right leaf. But prediction from `feval()` returns the result of the left leaf (what is totally wrong with zero value of the feature). Is there any extra tree transformation after feval function done?

**Comment by guolinke** [comment]
@qashqay654 
are the values of this feature all zeros? 
Can you give me the unique value of this feature?

**Comment by qashqay654** [comment]
@guolinke There are 130 thousand of unique values in this feature. Values are in the range from -1.3238272e+09 to 1.1378418e+09. Do you need to see them all?

**Comment by guolinke** [comment]
@qashqay654  can you try the ```test``` branch? only train one tree, then give me the log and the split point at tree model ?

**Comment by qashqay654** [comment]
@guolinke I think, I know, how you can repeat that problem. I'm using rather big datasets (10 million rows X 500 columns). But when I reduce dataset to 1000 rows X 10 columns, everything seems fine. So I've made simple code with random data, that gives described problem. Maybe that will help.

```
import lightgbm as lgb
import numpy as np

x_tr_sc = np.random.randn(1000000, 500)
x_te_sc = np.random.randn(10000, 500)
y_tr_sc = np.random.randn(1000000)
y_te_sc = np.random.randn(10000)

lgb_train = lgb.Dataset(x_tr_sc, y_tr_sc, free_raw_data=False, max_bin=1024)
lgb_eval = lgb.Dataset(x_te_sc, y_te_sc, free_raw_data=False, max_bin=1024)

params = {
    'task': 'train',
    'boosting_type': 'gbdt',
    'objective': 'regression_l2',
    'metric': ['l2'],
    'num_leaves': 31,
    'learning_rate': 0.3,
    'feature_fraction': 0.7,
    'bagging_fraction': 0.9,
    'bagging_freq': 5,
    'verbose': 0,
    'num_threads': 10,
    'min_data_in_leaf': 100
}

def feval(preds, train_data):
    pass
    return 'metric', 0, True

evals_result = {}
gbm = lgb.train(
    params,
    lgb_train,
    num_boost_round=5,
    valid_sets=[lgb_train, lgb_eval],
    feval=feval,
    evals_result=evals_result)

prediction = gbm.predict(x_te_sc)
pred_leaf = gbm.predict(x_te_sc, pred_leaf=True)

print max((prediction - gbm._Booster__inner_predict_buffer[1]) / prediction)

global_pred = gbm._Booster__inner_predict_buffer[1]

print np.where(np.abs(global_pred - prediction) >= 1e-8)[0][:10]
```

**Comment by guolinke** [comment]
@qashqay654 this code can reproduce the problem?

**Comment by qashqay654** [comment]
@guolinke yes, I see a lot of incorrectly predicted nodes.

**Comment by qashqay654** [comment]
@guolinke interesting point: this bug occurs only in v2. Branch v1 works well.

**Comment by guolinke** [comment]
@qashqay654 
fixed, the reason is a typo ....
